snuba/datasets/discover.py

```diff
@@ -115,10 +106,31 @@ def match_query_to_table(
                     if event_type == "transaction":
                         return EVENTS
 
-    if len(event_types) == 1 and event_types.pop() == "transaction":
+    if len(event_types) == 1 and "transaction" in event_types:
         return TRANSACTIONS
 
-    return EVENTS
+    if len(event_types) > 0 and "transaction" not in event_types:
+        return EVENTS
+
+    # If we cannot clearly pick a table from the top level conditions, then
+    # inspect the columns requested to infer a selection.
+    has_event_columns = False
+    has_transaction_columns = False
+    for col in query.get_all_ast_referenced_columns():
+        if events_only_columns.get(col.column_name):
+            has_event_columns = True
+        elif transactions_only_columns.get(col.column_name):
+            has_transaction_columns = True
+
+    if has_event_columns and has_transaction_columns:
+        # Impossible query, use the merge table
```

Comment:
Just a thought: This would benefit more refined logic. We should be striving to get as much data as possible back to the user. It might be worth it to do extra processing to determine the best table. For example, if I query for 5 transaction_only columns and 1 event_only column, we should use the transaction table. I'm not sure why all impossible queries should go to the events table.

Reply:
Yeah - I think "as much data as possible" has 2 possible and potentially conflicting interpretations here though - getting potentially all of the rows requested, or all of the columns requested - we can't guarantee both in this particular scenario. If we always pick today's events table - it means we are guaranteed to get all relevant rows, but not necessarily columns of transactions filled in. If we picked transactions or the future errors table we may get more of the columns but be missing rows. I assume that's why all "impossible" queries go to events today.

Reply:
Not all impossible queries go to events today. If the user has `type = transaction`, that will push an impossible query to use the transactions table instead.

I see your point about columns vs rows though. I feel pretty strongly that we should prioritize columns over rows, but I could be persuaded otherwise. It may also be that these kind of questions might be better answered by Sentry if/when this logic moves there.

Reply:
This seems to me more of a product decision given there is no `right` solution. What about taking this topic separately from this PR which is meant to make the table detection work after the tables split ?